clio-server/src/main/scala/org/broadinstitute/clio/server/service/IndexService.scala

```diff
@@ -67,4 +99,58 @@ abstract class IndexService[CI <: ClioIndex](
       elasticsearchIndex
     )
   }
+
+  def validateUpsert(
+    existingMetadata: Json,
+    newMetadata: Json
+  ): Either[Exception, Json] = {
+    val differences = JsonDiff.diff(existingMetadata, newMetadata, remember = true)
+
+    if (differences.ops.nonEmpty) {
+      Right(newMetadata)
+    } else {
+      val diffs: String = differences.ops.map { op =>
+        op.toString
+      }.mkString("\n")
+      Left(
+        UpsertValidationException(
+          s"""Adding this document will overwrite the following existing metadata:
+               | $diffs. Use 'force=true' to overwrite the existing data.""".stripMargin
+        )
+      )
+    }
+  }
+
+  def queryMetadataForKey(
+    indexKey: clioIndex.KeyType
+  ): Source[Json, NotUsed] = {
+    val keyJson = indexKey.asJsonObject
+    val fieldsToDrop = keyJson.keys.toSeq
+
+    searchService
+      .simpleStringQueryMetadata(indexKey.asJson.toString())(
+        elasticsearchIndex,
+        queryConverter
+      )
+      .fold[Either[Exception, Option[Json]]](Right(None)) { (acc, storedDocs) =>
+        acc match {
```

Comment:
Since the `Left` case just leaves `acc` as-is, we could
```scala
acc.flatMap {
  case None => ???
  case Some(_) => ???
}
```
to save some boilerplate